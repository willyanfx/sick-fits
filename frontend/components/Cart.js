import styled from 'styled-components';
import calcTotalPrice from '../lib/calcTotalPrice';
import { useCart } from '../lib/cartState';
import formatMoney from '../lib/formatMoney';
import CartStyles from './styles/CartStyles';
import CloseButton from './styles/CloseButton';
import Supreme from './styles/Supreme';
import { useUser } from './User';

const CartItemStyles = styled.li`
  padding: 1rem 0;
  border-bottom: 1px solid var(--lightGrey);
  display: grid;
  grid-template-columns: auto 1fr auto;
  img {
    margin-right: 1rem;
    width: 100px;
  }

  h3,
  p {
    margin: 0;
  }
`;

const CartItem = ({ cartItem }) => (
  <CartItemStyles>
    <img
      src={cartItem.product.photo.image.publicUrlTransformed}
      alt={cartItem.product.name}
    />
    <div>
      <h3>{cartItem.product.name}</h3>
      <p>
        {formatMoney(cartItem.product.price * cartItem.quantity)} -
        <em>
          {cartItem.quantity} &times; {formatMoney(cartItem.product.price)} each
        </em>
      </p>
    </div>
  </CartItemStyles>
);
function Cart() {
  const me = useUser();
  const { cartOpen, closeCart } = useCart();

  if (!me) return null;

  return (
    <CartStyles open={cartOpen}>
      <header>
        <Supreme>{me.name}'s Cart</Supreme>
        <CloseButton type="button" onClick={closeCart}>
          &times;
        </CloseButton>
      </header>
      <ul>
        {me.cart.map((cartItem) => (
          <CartItem key={cartItem.id} cartItem={cartItem} />
        ))}
      </ul>
      <footer>
        <p>{formatMoney(calcTotalPrice(me.cart))}</p>
      </footer>
    </CartStyles>
  );
}

export default Cart;
