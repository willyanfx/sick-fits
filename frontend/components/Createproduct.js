import { useMutation } from '@apollo/client';
import gql from 'graphql-tag';
import Router from 'next/router';
import useForm from '../lib/useForm';
import Form from './styles/Form';
import ErrorMessage from './ErrorMessage';
import { ALL_PRODUCTS_QUERY } from './Products';

const CREATE_PRODUCT_MUTATION = gql`
  mutation CREATE_PRODUCT_MUTATION(
    # variables passed and types
    $name: String!
    $description: String!
    $price: Int!
    $image: Upload
  ) {
    createProduct(
      data: {
        name: $name
        description: $description
        price: $price
        status: "AVAILABLE"
        photo: { create: { image: $image, altText: $name } }
      }
    ) {
      id
      name
      price
      description
    }
  }
`;

export default function CreateProduct() {
  const { inputs, handleChange, resetForm, clearForm } = useForm({});

  const [createProduct, { loading, error, data }] = useMutation(
    CREATE_PRODUCT_MUTATION,
    {
      variables: inputs,
      refetchQueries: [{ query: ALL_PRODUCTS_QUERY }],
    }
  );

  const handleSubmit = async (e) => {
    e.preventDefault();
    const res = await createProduct();
    clearForm();

    Router.push({
      pathname: `/product/${res.data.createProduct.id}`,
    });
  };

  return (
    <Form onSubmit={handleSubmit}>
      <ErrorMessage error={error} />
      <fieldset aria-busy={loading} disabled={loading}>
        <label htmlFor="file">
          Image
          <input
            type="file"
            id="image"
            name="image"
            value={inputs.file}
            onChange={handleChange}
            required
          />
        </label>
        <label htmlFor="name">
          Name
          <input
            type="text"
            id="name"
            name="name"
            placeholder="name"
            value={inputs.name}
            onChange={handleChange}
            required
          />
        </label>
        <label htmlFor="price">
          Price
          <input
            type="number"
            id="price"
            name="price"
            placeholder="price"
            value={inputs.price}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="description">
          Description
          <textarea
            id="description"
            name="description"
            placeholder="description"
            value={inputs.description}
            onChange={handleChange}
          />
        </label>
        <button type="submit">+ Add Product</button>
      </fieldset>
    </Form>
  );
}
