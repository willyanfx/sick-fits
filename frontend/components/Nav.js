import Link from 'next/link';

export default function Nav() {
  return (
    <nav>
      <Link href="/products">Products</Link>
      <Link href="/sell">Sell</Link>
      <Link href="/orders">Orders</Link>
    </nav>
  );
}
