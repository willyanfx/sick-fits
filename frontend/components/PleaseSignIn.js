import { useUser } from './User';
import SignIn from './SignIn';

function PleaseSignIn({ children }) {
  const me = useUser();

  if (!me) {
    return <SignIn />;
  }

  return children;
}

export default PleaseSignIn;
