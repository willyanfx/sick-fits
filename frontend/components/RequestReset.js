import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import Form from './styles/Form';
import useForm from '../lib/useForm';
import { CURRENT_USER_QUERY } from './User';
import ErrorMessage from './ErrorMessage';

export const REQUEST_RESET_MUTATION = gql`
  mutation REQUEST_RESET_MUTATION($email: String!) {
    sendUserPasswordResetLink(email: $email) {
      message
      code
    }
  }
`;

export default function RequestReset() {
  const { inputs, handleChange, resetForm } = useForm({
    email: '',
  });
  const [signup, { data, loading }] = useMutation(REQUEST_RESET_MUTATION, {
    variables: inputs,
  });

  const handleSubmit = async (e) => {
    e.preventDefault();
    console.log(inputs);
    const res = await signup();
    console.log('RESULT::: ', res);
    resetForm();
  };
  console.log(data?.authenticateUserWithPassword?.message);

  const resetedLink = data?.sendUserPasswordResetLink === null && (
    <p>Success! check your email for a link</p>
  );
  return (
    <Form method="POST" onSubmit={handleSubmit}>
      <h2>Request a Password Reset</h2>
      {loading && <p>Loading ...</p>}
      <ErrorMessage error={data?.authenticateUserWithPassword} />
      <fieldset aria-busy={loading} disabled={loading}>
        {resetedLink}

        <label htmlFor="email">
          Email
          <input
            type="email"
            name="email"
            autoComplete="email"
            placeholder="Your email address"
            value={inputs.email}
            onChange={handleChange}
          />
        </label>

        <button type="submit">Reset Password</button>
      </fieldset>
    </Form>
  );
}
