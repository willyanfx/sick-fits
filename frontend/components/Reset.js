import gql from 'graphql-tag';
import { useMutation } from '@apollo/client';
import Form from './styles/Form';
import useForm from '../lib/useForm';
import { CURRENT_USER_QUERY } from './User';
import ErrorMessage from './ErrorMessage';

const RESET_MUTATION = gql`
  mutation RESET_MUTATION(
    $email: String!
    $token: String!
    $password: String!
  ) {
    redeemUserPasswordResetToken(
      email: $email
      token: $token
      password: $password
    ) {
      message
      code
    }
  }
`;

function Reset({ token }) {
  const { inputs, handleChange, resetForm } = useForm({
    email: '',
    password: '',
    token,
  });
  const [reset, { data, loading }] = useMutation(RESET_MUTATION, {
    variables: inputs,
  });

  const handleSubmit = async (e) => {
    e.preventDefault();
    await reset();
    resetForm();
  };
  console.log(data?.authenticateUserWithPassword?.message);

  const resetedLink = data?.redeemUserPasswordResetToken === null && (
    <p>Success! You can sign</p>
  );
  return (
    <Form method="POST" onSubmit={handleSubmit}>
      <h2>Reset Your Password</h2>
      {loading && <p>Loading ...</p>}
      <ErrorMessage error={data?.redeemUserPasswordResetToken} />
      <fieldset aria-busy={loading} disabled={loading}>
        {resetedLink}

        <label htmlFor="email">
          Email
          <input
            type="email"
            name="email"
            autoComplete="email"
            placeholder="Your email address"
            value={inputs.email}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="password">
          Password
          <input
            type="password"
            name="password"
            placeholder="password"
            value={inputs.password}
            onChange={handleChange}
          />
        </label>

        <button type="submit">Reset Password</button>
      </fieldset>
    </Form>
  );
}
export { Reset };
