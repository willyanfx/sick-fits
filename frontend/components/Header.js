import Link from 'next/link';
import Nav from './Nav';

export default function Header() {
  return (
    <header>
      <div className="bar">
        Sick fits
        <Link href="/">Sick Fits</Link>
      </div>
      <div className="sub-bar">
        <p>search</p>
      </div>
      <Nav />
    </header>
  );
}
