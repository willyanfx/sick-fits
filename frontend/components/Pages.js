import PropTypes from 'prop-types';
import Header from './Header';

export default function Page({ children }) {
  return (
    <div>
      <Header />
      <h2>I am page component</h2>
      <p>{children}</p>
    </div>
  );
}

Page.propTypes = {
  children: PropTypes.any,
};
