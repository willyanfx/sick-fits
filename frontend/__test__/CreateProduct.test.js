import { render, screen, waitFor } from '@testing-library/react';
import { MockedProvider } from '@apollo/client/testing';
import userEvent from '@testing-library/user-event';
import Router from 'next/router'; // We will MOCK THIS
import wait from 'waait';
import CreateProduct, {
  CREATE_PRODUCT_MUTATION,
} from '../components/Createproduct';
import { fakeItem, makePaginationMocksFor } from '../lib/testUtils';
import { ALL_PRODUCTS_QUERY } from '../components/Products';

const item = fakeItem();

jest.mock('next/router', () => ({
  push: jest.fn(),
}));

describe('<CreateProduct/>', () => {
  it('should render and matches snapshot', () => {
    const { container, debug } = render(
      <MockedProvider>
        <CreateProduct />
      </MockedProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('should handles the updating', async () => {
    const { container, debug } = render(
      <MockedProvider>
        <CreateProduct />
      </MockedProvider>
    );

    await userEvent.type(screen.getByPlaceholderText(/Name/i), item.name);
    await userEvent.type(
      screen.getByPlaceholderText(/Price/i),
      item.price.toString()
    );
    await userEvent.type(
      screen.getByPlaceholderText(/Description/i),
      item.description
    );

    expect(screen.getByDisplayValue(item.name)).toBeInTheDocument();
    expect(screen.getByDisplayValue(item.description)).toBeInTheDocument();
    expect(screen.getByDisplayValue(item.price)).toBeInTheDocument();
  });

  it('should creates the items when the form is submitted', async () => {
    // create the mocks for this one
    const mocks = [
      {
        request: {
          query: CREATE_PRODUCT_MUTATION,
          variables: {
            name: item.name,
            description: item.description,
            image: '',
            price: item.price,
          },
        },
        result: {
          data: {
            createProduct: {
              ...item, // all fake item fields
              id: 'abc123',
            },
          },
        },
      },
      {
        request: {
          query: ALL_PRODUCTS_QUERY,
          variables: { skip: 0, first: 2 },
        },
        result: {
          data: {
            allProducts: [item],
          },
        },
      },
    ];

    const { container, debug } = render(
      <MockedProvider mocks={mocks}>
        <CreateProduct />
      </MockedProvider>
    );

    userEvent.type(screen.getByPlaceholderText(/Name/i), item.name);
    userEvent.type(screen.getByPlaceholderText(/Price/i), String(item.price));
    userEvent.type(
      screen.getByPlaceholderText(/Description/i),
      item.description
    );

    userEvent.click(screen.getByText(/Add Product/));
    await waitFor(() => wait(0));
    expect(Router.push).toHaveBeenCalled();
    expect(Router.push).toHaveBeenCalledWith({ pathname: '/product/abc123' });
  });
});
