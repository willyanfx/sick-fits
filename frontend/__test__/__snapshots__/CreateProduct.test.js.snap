// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreateProduct/> should render and matches snapshot 1`] = `
<div>
  <form
    class="Form-sc-1xszr8q-0 JYcwP"
  >
    <fieldset
      aria-busy="false"
    >
      <label
        for="file"
      >
        Image
        <input
          id="image"
          name="image"
          required=""
          type="file"
          value=""
        />
      </label>
      <label
        for="name"
      >
        Name
        <input
          id="name"
          name="name"
          placeholder="name"
          required=""
          type="text"
          value=""
        />
      </label>
      <label
        for="price"
      >
        Price
        <input
          id="price"
          name="price"
          placeholder="price"
          type="number"
          value="0"
        />
      </label>
      <label
        for="description"
      >
        Description
        <textarea
          id="description"
          name="description"
          placeholder="description"
        />
      </label>
      <button
        type="submit"
      >
        + Add Product
      </button>
    </fieldset>
  </form>
</div>
`;
